File: lib/vaultPathValidator.py
"""
Vault Path Security Validator & Traversal Prevention Module
"""

import re
from pathlib import Path
from typing import Union

class VaultPathValidator:
    """Whitelist vault paths and reject path traversal attacks."""

    PATH_PATTERN = re.compile(r"^[a-zA-Z0-9_\-\.]+$")

    @classmethod
    def validate_vault_id(cls, vault_id: str) -> bool:
        """Reject traversal patterns: ../../etc/passwd, /etc/passwd, $(whoami)"""
        if not vault_id or not isinstance(vault_id, str):
            raise ValueError(f"Invalid vault_id type: {type(vault_id)}")
        if ".." in vault_id or "/" in vault_id or "\\" in vault_id:
            raise ValueError(f"Path traversal detected in vault identifier: {vault_id}")
        if not cls.PATH_PATTERN.match(vault_id):
            raise ValueError(f"Disallowed characters in vault identifier: {vault_id}")
        return True

    @classmethod
    def secure_vault_path(cls, root_dir: Union[str, Path], vault_id: str) -> Path:
        """Build path safely, ensuring no escape from allowed root directory."""
        cls.validate_vault_id(vault_id)
        root = Path(root_dir).resolve()
        target = (root / vault_id).resolve()

        # Enforce boundary confinement
        try:
            target.relative_to(root)
        except ValueError:
            raise PermissionError(f"Path traversal detected: {vault_id} resolves outside root {root}")

        return target

File: lib/vaultPathValidator.mjs
/**
 * Vault Path Security Validator (ESM)
 * Prevents directory traversal attacks in Node.js endpoints and file handlers.
 */

import path from 'node:path';

const PATH_PATTERN = /^[a-zA-Z0-9_\-\.]+$/;

export class VaultPathValidator {
  static validateVaultId(vaultId) {
    if (!vaultId || typeof vaultId !== 'string') {
      throw new TypeError(`Invalid vaultId: ${vaultId}`);
    }
    if (vaultId.includes('..') || vaultId.includes('/') || vaultId.includes('\\')) {
      throw new Error(`Path traversal sequence detected in vault identifier: ${vaultId}`);
    }
    if (!PATH_PATTERN.test(vaultId)) {
      throw new Error(`Disallowed characters in vault identifier: ${vaultId}`);
    }
    return true;
  }

  static secureVaultPath(rootDir, vaultId) {
    this.validateVaultId(vaultId);
    const resolvedRoot = path.resolve(rootDir);
    const targetPath = path.resolve(resolvedRoot, vaultId);

    if (!targetPath.startsWith(resolvedRoot)) {
      throw new Error(`Access Denied: Path escapes permitted root ${resolvedRoot}`);
    }
    return targetPath;
  }
}
